File: functions/src/payout-account.ts
import {HttpsError} from "firebase-functions/v2/https";
import * as logger from "firebase-functions/logger";
import * as admin from "firebase-admin";

/**
 * Creating a Razorpay Route linked account for a pro or host.
 *
 * **Raw bank details and PAN never touch Firestore.** They arrive in the
 * callable, go straight to Razorpay, and only the returned account id plus a
 * masked last-4 is persisted. Storing an account number would make Firestore a
 * far more attractive target than it already is, for no product benefit — the
 * partner never needs to read it back, and Razorpay is the system of record.
 *
 * Nothing is ever logged from the payload for the same reason.
 */

export type BankDetails = {
  name: string;
  email: string;
  pan: string;
  accountNumber: string;
  ifsc: string;
  beneficiaryName: string;
};

const IFSC_RE = /^[A-Z]{4}0[A-Z0-9]{6}$/;
const PAN_RE = /^[A-Z]{5}[0-9]{4}[A-Z]$/;

export function readBankDetails(data: unknown): BankDetails {
  const d = (data ?? {}) as Record<string, unknown>;
  const s = (k: string) => String(d[k] ?? "").trim();

  const details: BankDetails = {
    name: s("name"),
    email: s("email").toLowerCase(),
    pan: s("pan").toUpperCase(),
    accountNumber: s("accountNumber").replace(/\s/g, ""),
    ifsc: s("ifsc").toUpperCase(),
    beneficiaryName: s("beneficiaryName"),
  };

  // Validated here rather than only in the app: a callable is reachable without
  // the app, and a malformed account number means money going nowhere.
  if (!details.name) throw new HttpsError("invalid-argument", "name-required");
  if (!details.email.includes("@")) throw new HttpsError("invalid-argument", "email-invalid");
  if (!PAN_RE.test(details.pan)) throw new HttpsError("invalid-argument", "pan-invalid");
  if (!IFSC_RE.test(details.ifsc)) throw new HttpsError("invalid-argument", "ifsc-invalid");
  if (details.accountNumber.length < 6 || !/^\d+$/.test(details.accountNumber)) {
    throw new HttpsError("invalid-argument", "account-number-invalid");
  }
  if (!details.beneficiaryName) {
    throw new HttpsError("invalid-argument", "beneficiary-name-required");
  }
  return details;
}

type RazorpayRoute = {
  accounts: {create: (p: object) => Promise<{id: string}>};
  stakeholders: {create: (accountId: string, p: object) => Promise<{id: string}>};
  products: {
    requestProductConfiguration: (accountId: string, p: object) => Promise<{id: string}>;
    edit: (accountId: string, productId: string, p: object) => Promise<unknown>;
  };
};

/**
 * Route onboarding is three calls: the account, a stakeholder (the person
 * behind it, with their PAN), then the `route` product configured with the
 * settlement bank account. Razorpay then runs its own checks — the account is
 * not immediately payable, which is why status starts as `pending`.
 */
export async function createLinkedAccount(
  rzp: RazorpayRoute,
  uid: string,
  d: BankDetails,
): Promise<{accountId: string}> {
  const account = await rzp.accounts.create({
    email: d.email,
    phone: "",
    type: "route",
    legal_business_name: d.name,
    business_type: "individual",
    contact_name: d.name,
    profile: {category: "services", subcategory: "other"},
    reference_id: uid, // ties the Razorpay account back to the Pawgo user
  });

  await rzp.stakeholders.create(account.id, {
    name: d.beneficiaryName,
    email: d.email,
    kyc: {pan: d.pan},
  });

  const product = await rzp.products.requestProductConfiguration(account.id, {
    product_name: "route",
    tnc_accepted: true,
  });

  await rzp.products.edit(account.id, product.id, {
    settlements: {
      account_number: d.accountNumber,
      ifsc_code: d.ifsc,
      beneficiary_name: d.beneficiaryName,
    },
    tnc_accepted: true,
  });

  return {accountId: account.id};
}

/** Persists only what is safe to keep. */
export async function savePayoutAccount(uid: string, accountId: string, d: BankDetails) {
  await admin.firestore().collection("payoutAccounts").doc(uid).set({
    uid,
    razorpayAccountId: accountId,
    // Enough for a partner to recognise which account they gave us, and
    // useless to anyone who steals it.
    bankLast4: d.accountNumber.slice(-4),
    ifsc: d.ifsc,
    beneficiaryName: d.beneficiaryName,
    status: "pending", // Razorpay has to run its own checks before it is payable
    createdAt: Date.now(),
    updatedAt: Date.now(),
    error: "",
  });
  logger.info("payout account created", {uid, accountId});
}

File: lib/data/models/payout_account.dart
enum PayoutAccountStatus {
  pending('pending'),
  active('active'),
  failed('failed');

  final String storageKey;
  const PayoutAccountStatus(this.storageKey);

  static PayoutAccountStatus fromStorage(String key) => PayoutAccountStatus.values
      .firstWhere((s) => s.storageKey == key, orElse: () => PayoutAccountStatus.pending);
}

/// What Pawgo keeps about a partner's bank account — deliberately almost
/// nothing.
///
/// The account number and PAN are sent straight to Razorpay by the
/// `createPayoutAccount` Function and are **never written to Firestore**.
/// Razorpay is the system of record; the partner never needs to read them back,
/// so storing them would add risk for no product benefit. [bankLast4] exists
/// only so someone can recognise which account they gave us.
class PayoutAccount {
  final String uid, razorpayAccountId, bankLast4, ifsc, beneficiaryName, error;
  final PayoutAccountStatus status;

  const PayoutAccount({
    required this.uid,
    required this.razorpayAccountId,
    required this.bankLast4,
    required this.ifsc,
    required this.beneficiaryName,
    this.status = PayoutAccountStatus.pending,
    this.error = '',
  });

  factory PayoutAccount.fromMap(String uid, Map<String, dynamic> m) => PayoutAccount(
        uid: uid,
        razorpayAccountId: (m['razorpayAccountId'] ?? '') as String,
        bankLast4: (m['bankLast4'] ?? '') as String,
        ifsc: (m['ifsc'] ?? '') as String,
        beneficiaryName: (m['beneficiaryName'] ?? '') as String,
        status: PayoutAccountStatus.fromStorage((m['status'] ?? 'pending') as String),
        error: (m['error'] ?? '') as String,
      );
}

enum PayoutStatus {
  owed('owed', 'Owed'),
  held('held', 'Processing'),
  released('released', 'Paid out'),
  reversed('reversed', 'Cancelled'),
  failed('failed', 'Failed');

  final String storageKey, label;
  const PayoutStatus(this.storageKey, this.label);

  static PayoutStatus fromStorage(String key) => PayoutStatus.values
      .firstWhere((s) => s.storageKey == key, orElse: () => PayoutStatus.owed);
}

/// One booking's worth of earnings. Written by the server when a customer pays;
/// clients only ever read their own.
class Payout {
  final String id, kind, bookingId, partnerId;
  final int amount, dueAt, createdAt;
  final PayoutStatus status;

  const Payout({
    required this.id,
    required this.kind,
    required this.bookingId,
    required this.partnerId,
    required this.amount,
    required this.status,
    this.dueAt = 0,
    this.createdAt = 0,
  });

  factory Payout.fromMap(String id, Map<String, dynamic> m) => Payout(
        id: id,
        kind: (m['kind'] ?? '') as String,
        bookingId: (m['bookingId'] ?? '') as String,
        partnerId: (m['partnerId'] ?? '') as String,
        amount: (m['amount'] ?? 0) as int,
        status: PayoutStatus.fromStorage((m['status'] ?? 'owed') as String),
        dueAt: (m['dueAt'] ?? 0) as int,
        createdAt: (m['createdAt'] ?? 0) as int,
      );
}

/// Totals a partner sees. `pending` covers everything not yet paid out —
/// reversed payouts are excluded because that money went back to the customer.
class EarningsSummary {
  final int pending, paid;
  const EarningsSummary({this.pending = 0, this.paid = 0});

  factory EarningsSummary.from(List<Payout> payouts) {
    var pending = 0;
    var paid = 0;
    for (final p in payouts) {
      switch (p.status) {
        case PayoutStatus.owed:
        case PayoutStatus.held:
          pending += p.amount;
        case PayoutStatus.released:
          paid += p.amount;
        case PayoutStatus.reversed:
        case PayoutStatus.failed:
          break;
      }
    }
    return EarningsSummary(pending: pending, paid: paid);
  }
}
